### ISSUE TITLE
Running behind nginx - update subpath in the cross-references in the app?

### ISSUE THREAD

**Issue by svilupp** [GenieFramework/Stipple.jl#199]
How do I change the app's own references if it's running behind an nginx proxy?

For example, let's say my nginx.conf has the following route:
```
location /demoapp/ {
            proxy_pass http://127.0.0.1:8000/;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header Host $host;
            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection "upgrade";
        }
```

How would I change the Stipple configuration, so that all its cross-references honour the `/demoapp/` subpath?
Eg, some of the assets like `<link href="/stipple.jl/master/assets/css/stipplecore.css" rel="stylesheet">` would be linking to 
`<link href="/pricing/stipple.jl/master/assets/css/stipplecore.css" rel="stylesheet">` instead (and the routes would be updated as well)

I can see they are pushed from [here](https://github.com/GenieFramework/Stipple.jl/blob/aacd3065445111fb44229d95197ee7a4f7e01f1f/src/Stipple.jl#L525)

Which would mean that perhaps what I need to change is the [`AssetConfig`](https://github.com/GenieFramework/Genie.jl/blob/46aa5bad5eebab75427b422a86ad3514d89bcebd/src/Assets.jl#L72)?
(and overwrite the [default](https://github.com/GenieFramework/Stipple.jl/blob/aacd3065445111fb44229d95197ee7a4f7e01f1f/src/stipple/assets.jl#LL12C28-L12C28))

I'm aware of the Genie.config parameter `server_document_root`, but that seems to be only for the Genie assets (and about the local storage).

*** 
Is it perhaps easier to explicitly define all the assets to be forwarded in nginx? (It wouldn't solve if you want to move between routes within your app)

**Comment by hhaensel** [comment]
I remember having used the parameter 'basepath' in the past.

**Comment by essenciary** [comment]
That's right. Either set `Genie.config.base_path` or set a `BASEPATH` env variable (the env will overwrite the config if both are set).

**Comment by svilupp** [comment]
It seems to be exactly what I need, but I'm struggling to have it reflected in the asset paths when testing it locally (eg, `<link href="/stipple.jl/master/assets/css/stipplecore.css" rel="stylesheet">`)

I'll test it on nginx tomorrow to see if it's any different.

Am I meant to do anything except for:
```
Genie.Configuration.config!(; app_env = "dev", base_path = "/pricing/")
Server.isrunning() || Server.up(8000)
```
?

I can see that it's properly reflected in config, but not anywhere else.

**Comment by essenciary** [comment]
Nginx should not make a difference, it should be reflected here... What's the structure of the app? Can you show a tree of the files and folders? The environment and the configuration are loaded earlier in the starting cycle of the app so possibly they're already set. You can confirm that by passing the `BASEPATH` env to the julia/genie process (depending on how you load the app). If that works then it needs to be loaded earlier.

**Comment by svilupp** [comment]
Great point! You were right. 

My config! call was right after the `@genietools`, which was already too late because of [this line](https://github.com/GenieFramework/GenieFramework.jl/blob/e0e6d400b08782cf9f6860c75218bf691774f4ee/src/GenieFramework.jl#L56)

I can confirm the BASEPATH works well, producing asset links like: "<script src="/demoapp/stipple.jl/master/assets/js/main_reactivemodel.js" defer=""></script>"

**Comment by essenciary** [comment]
> My config! call was right after the @genietools, which was already too late because of [this line](https://github.com/GenieFramework/GenieFramework.jl/blob/e0e6d400b08782cf9f6860c75218bf691774f4ee/src/GenieFramework.jl#L56).
Moreover, this line also means that config! cannot be used (is it a bug or by design? I'll open an issue to flag it)

Yes, it should be definitely reviewed. 

The way I routinely use the two options (`ENV` and `config!`) are: 
* `config!` with dedicated per environment configuration files (in `config/env/*.jl`). These provide for one single place for storing configuration and they are loaded by Genie at the right time. 
* for Docker deployments using `ENV` 

In your use case, you're setting a config and then starting the server. Which means that the server start part of the framework should actually recheck the configuration. We're already doing that for some configurations (port, host and ws_port) which can also be passed as arguments to `Server.up`. I think the easiest would be to also add similar support for `base_path`.
